File: 2020/deno/seven.ts
const input = await Deno.readTextFile("./input/seven.in").then((res) =>
  res.split("\n")
);

/**
 * Helpers
 */

const MY_BAG = "shiny gold";

const processBag = (content: string) => {
  if (content.includes("no other")) return {};

  const [qty, color, accent] = content.split(" ");

  return { [`${color} ${accent}`]: parseInt(qty) };
};

/**
 * Part One
 */

interface Bag extends Record<string, string | number> {
  name: string;
}

const bags: Bag[] = input.map((row) => {
  const [fullColor, spec] = row.split("contain");
  const [accent, color] = fullColor.split(" ");

  const contents = spec
    .split(",")
    .map((x) => x.trim())
    .reduce<Record<string, number>>((prev, curr) => {
      return {
        ...prev,
        ...processBag(curr)
      };
    }, {});

  return {
    ...contents,
    name: `${accent} ${color}`
  };
});

const container = new Set(
  bags.filter((x) => x[MY_BAG]).map(({ name }) => name)
);

let prevSize = 0;

while (1) {
  prevSize = container.size;

  for (const entry of [...container]) {
    bags
      .filter((bag) => bag[entry])
      .flat(Infinity)
      .forEach(({ name }) => container.add(name));
  }

  if (container.size === prevSize) break;
}

console.log("Part One:", container.size);

/**
 * Part Two
 */

const countNested = (key: string, countSelf: boolean = false): number => {
  const current = bags.find((x) => x.name === key);

  if (!current) throw new Error(`Bag not found: ${key}`);

  if (Object.keys(current).length === 1) return 0;

  return Object.keys(current)
    .filter((x) => x !== "name")
    .map((x) => {
      const qty = current[x] as number;

      const mult = countNested(x, true);

      if (!mult) {
        return qty;
      }

      return mult * qty;
    })
    .reduce((prev, curr) => prev + curr, countSelf ? 1 : 0);
};

console.log("Part Two:", countNested(MY_BAG));

File: 2020/seven/src/bin/seven.rs
extern crate seven;
use seven::{part_one, part_two};
use std::fs;

fn main() {
    let raw_input = fs::read_to_string("./input/seven.in").expect("Error reading input");
    let input = raw_input
        .split_terminator("\n")
        .map(|x| x.to_string())
        .map(|entry| {
            entry
                .split_terminator(" contain ")
                .map(|x| {
                    x.trim()
                        .to_string()
                        .replace("bags", "")
                        .replace("bag", "")
                        .replace(".", "")
                })
                .collect()
        })
        .collect::<Vec<Vec<String>>>();

    println!("Part One: {}", part_one(&input));
    println!("Part Two: {}", part_two(&input));
}
